File: Queue/C/CirQueue.c
#include<stdio.h>
#include<stdlib.h>

struct node
{
    int data;
    struct node * next;
};

typedef struct node NODE;
typedef struct node * PNODE;
typedef struct node ** PPNODE;

void enqueue(PPNODE first, PPNODE last, int no)
{
    PNODE newn = NULL;

    newn = (PNODE)malloc(sizeof(NODE));
    newn->data = no;
    newn->next = NULL;

    if((*first == NULL) && (*last == NULL))
    {
        *first = newn;
        *last = newn;

        (*last)->next = *first;
    }
    else
    {
        (*last)->next = newn;
        *last = newn;

        (*last)->next = *first;
    }
}

int dequeue(PPNODE first, PPNODE last)
{
    PNODE temp = NULL;
    int value = 0;

    if((*first == NULL) && (*last == NULL))
    {
        printf("Unable to delete Queue is empty\n");
        return -1;
    }
    else if(*first == *last)
    {
        value = (*first)->data;

        free(*first);

        *first = NULL;
        *last = NULL;
    }
    else
    {
        temp = *first;
        value = temp->data;

        *first = (*first)->next;
        (*last)->next = *first;
        
        free(temp);
    }

    return value;
}

void Display(PNODE first, PNODE last)
{
    if((first == NULL) && (last == NULL))
    {
        printf("Queue is empty\n");
        return;
    }

    do
    {
        printf(" | %d | - ",first->data);
        first = first->next;

    } while(first != last->next);

    printf("\n");
    
}

int Count(PNODE first, PNODE last)
{
    int iCount = 0;

    if((first == NULL) && (last == NULL))
    {
        printf("Queue is empty\n");
        return 0;
    }

    do
    {
        iCount++;
        first = first->next;

    } while(first != last->next);

    return iCount;
}

int main()
{
    PNODE head = NULL;
    PNODE tail = NULL;
    int iRet = 0;
    int choice = 0;
    int value = 0;

    printf("\n=============================================================\n");
    printf("                       Circular Queue Manager                    \n");
    printf("=============================================================\n");

    while(1)
    {
        printf("\n-------------------------------------------------------------\n");
        printf("Please select an operation:\n");
        printf("-------------------------------------------------------------\n");
        printf(" 1 : Enqueue at rear\n");
        printf(" 2 : Dequeue from front\n");
        printf(" 3 : Display from front to rear\n");
        printf(" 4 : Count total elements\n");
        printf(" 0 : Exit\n");
        printf("-------------------------------------------------------------\n");

        printf("Enter your choice: \n");
        scanf("%d",&choice);

        if(choice == 1)
        {
            printf("Enter element to enqueue at rear: \n");
            scanf("%d",&value);
            enqueue(&head, &tail, value);
        }
        else if(choice == 2)
        {
            value = dequeue(&head, &tail);
            if(value != -1)
            {
                printf("Dequeued from front: %d\n",value);
            }
        }
        else if(choice == 3)
        {
            Display(head, tail);
        }
        else if(choice == 4)
        {
            iRet = Count(head, tail);
            printf("Total elements in queque: %d\n",iRet);
        }
        else if(choice == 0)
        {
            printf("Thank you for using our Application!\n");
            break;
        }
        else
        {
            printf("Invalid choice. Please try again.\n");
        }
    }

    return 0;
};

File: Queue/Python/CirQueue.py

class node:

    def __init__(self, no):
        self.data = no
        self.next = None

class CirQueue:

    def __init__(self):
        self.first = None
        self.last = None
        self.iCount = 0

    def enqueue(self, no):

        newn = node(no)

        if(self.first == None and self.last == None):
            self.first = newn
            self.last = newn

            self.last.next = self.first

        else:
            self.last.next = newn
            self.last = newn

            self.last.next = self.first

        self.iCount += 1

    def dequeue(self):

        if(self.first == None and self.last == None):
            print("Unable to delete Queue is empty")
            return None

        elif(self.first == self.last):
            iValue = self.first.data

            del self.first

            self.first = None
            self.last = None

        else:
            temp = self.first
            iValue = temp.data

            self.first = self.first.next
            self.last.next = self.first
            del temp

        self.iCount -= 1
        return iValue

    def Display(self):

        if(self.first == None and self.last == None):
            print("Queue is empty")
            return

        temp = self.first

        while(True):
            print(f" | {temp.data} | ->",end="")
            temp = temp.next

            if(temp == self.last.next):
                break

    def Count(self):
        return self.iCount

def main():

    obj = CirQueue()

    print("=============================================================");
    print("                       Circular Queue Manager                    ");
    print("=============================================================");

    while True:
        print("\n-------------------------------------------------------------")
        print("Please select an operation:")
        print("-------------------------------------------------------------")
        print(" 1 : Enqueue at rear")
        print(" 2 : Dequeue from front")
        print(" 3 : Display from front to rear")
        print(" 4 : Count total elements")
        print(" 0 : Exit")
        print("-------------------------------------------------------------")

        choice = int(input("Enter your choice : "))

        if choice == 1:
            value = int(input("Enter element to enqueue at rear: "))
            obj.enqueue(value)

        elif choice == 2:
            iRet = obj.dequeue()
            if iRet is not None:
                print(f"Dequeued from front: {iRet}")

        elif choice == 3:
            obj.Display()

        elif choice == 4:
            print(f"Total elements in Queue: {obj.Count()}")

        elif choice == 0:
            print("Thank you for using our Application!")
            break

        else:
            print("Invalid choice, please try again")


if __name__ == "__main__":
    main()